### ISSUE TITLE
maven_jar does not pull transitive dependencies

### ISSUE THREAD

**Issue by davidzchen** [bazelbuild/bazel#89]
When I attempted to [build the Dropwizard Example with Bazel](https://github.com/davidzchen/dropwizard-example-bazel) using `maven_jar` to pull the artifacts the project depends on, I got a number of compile time errors that seems to suggest that the transitive dependencies of the artifacts were not being pulled.

The documentation for `maven_jar`'s `exclusions` attribute has a note that mentions that `exclusions` is "Not yet implemented" but does not state whether pulling transitive dependencies is supported. Is pulling transitive dependencies with `maven_jar` currently not supported?

**Comment by damienmg** [comment]
It is not yet supported. Assigning to kchodorow for more explanations about future plans.

**Comment by kchodorow** [comment]
I don't think automatically downloading transitive dependencies will be supported for the foreseeable future, as it would make it difficult to ensure dependencies were all the same on different machines and it would make specifying a sha1 for every dependency awkward.  See https://docs.google.com/document/d/1LIF_CXwamK6MAkLDekS83TlqL50sS7GqL9ocfwcmAu8/edit# for more on the design.

**Comment by damienmg** [comment]
I don't think we should disregards doing so. We could restrict it to fully defined version dependency only, i.e. we pull a dependency only if its version is not specified by a version range. That would ensure reproducibility without having to specify sha1 (eventually with an option to force pulling even if we encounter a version range). In the future plan, there is actually a "flag" for that (IMHO this flag belongs to the maven_jar rule)

**Comment by davidzchen** [comment]
I agree with Damien that we should still implement pulling transitive dependencies. Given that all other major Java build systems (i.e. Ant/Ivy, Maven, Gradle, etc.) all pull transitive dependencies by default, not having this feature would be a blocker for migrating many Java-based projects from one of these other build systems to Bazel.

Doesn't Maven pom files specify the versions for the artifacts that a given JAR depends on?

**Comment by talios** [comment]
@davidzchen It would be good if Bazel _didn't_ repeat the same bugs that Maven has in this regard tho.  Maven's "compile" scope is essentially a "compile _and_ runtime" scope, which is subtly different to its "runtime" scope.

If project-A has several compile time dependencies - there IT'S dependencies to compile, not necessarily be consumed by project-B's compilation.

Pulling in transitives for distribution and/or unit-testing is a different thing.

**Comment by davidzchen** [comment]
While that makes sense in theory, unfortunately, it does not always work very well in practice. There are many Java projects that rely on jars to pull in transitive dependencies in order to compile successfully.

A concrete example of this is my attempts at compiling `dropwizard-example` with Bazel. I am unable to simply use the list of dependencies in the [Maven pom](https://github.com/dropwizard/dropwizard/blob/4ceee223cddbac4183828c1c0277b1f9ec102a75/dropwizard-example/pom.xml) because the code directly references classes in jars that are transitive dependencies of the ones listed in the Maven pom. For example, it assumes that `dropwizard-jersey` will pull in `jersey-server`, etc.

IIRC, the other Bazel build rules have the same behavior: if `java_library` `foo` depends on `bar` and `baz`, and `bar` also depends on `baz`, then `foo` only needs to add `bar` to its `deps` and `baz` will be included transitively. The same applies to the `cc_*` rules from what I have observed. Please correct me if I'm wrong.

In any case, I think we need a better story around pulling Maven jars transitively. Otherwise, migrating something as simple as `dropwizard-example` would require one to dig deep into the various transitive dependencies among the `dropwizard` jars it depends on in order to cover all of the transitive dependencies, which is very time consuming and may discourage adoption, unless covering all transitive dependencies this way yields some properties that provides very clear benefits for projects that do want to pull jars from Maven rather than checking them into source control.

@kchodorow, @damienmg - Is there a significant number of artifacts in repositories such as Maven Central that are lacking version info for dependencies?
@talios - I would also like to learn more about the issues that Maven's behavior for its compile and runtime scopes. Can you give some examples of how this has caused issues?

**Comment by sunyc** [comment]
Would it be possible to write a pom->bazel translation tool that
automatically generate relevant rules?

**Comment by talios** [comment]
On 6 Apr 2015, at 1:50, Yucong Sun wrote:

> Would it be possible to write a pom->bazel translation tool that
> automatically generate relevant rules?

Actually, depending on the approach one took it shouldn't be all that difficult - ish, and certainly wouldn't be foolproof.

One could make use of the new [1] polyglot maven support and write a converter, that simply writes out a WORKSPACE file with all the dependencies, and potentially generate basic BUILD files for src/main and src/test for each module ( assuming a multi-module build ).

I did some work on the original clojure polyglot implementation and been meaning to check out whats changed, might have to try find some time and take a look.

[1] https://github.com/takari/maven-polyglot


-- 
Mark Derricutt
http://www.theoryinpractice.net
http://www.chaliceofblood.net
http://plus.google.com/+MarkDerricutt
http://twitter.com/talios
http://facebook.com/mderricutt

**Comment by kchodorow** [comment]
See http://bazel.io/docs/external.html ("Transitive dependencies" section) for the current status of this.  The last step I'm planning to take before resolving this issue is to add a target to the package maven_jar generates so that you can specify @foo//jar:transitive (@foo//jar:all? @foo//jar:kitchen-sink? naming suggestions welcome) to depend on a jar + all of its transitive dependencies.

**Comment by cristipp** [comment]
@kchodorow Did anything happen with the @foo//jar:transitive idea? The closing commit doesn't include it.

Absent that, assuming I'm starting a bazel project from scratch, what is the recommended flow? Should I write a pom.xml file, then run bazel run //src/tools/generate_workspace on top of it to generate a correct WORKSPACE, then depend on "transitive-deps"?

Maven dependencies are critical for Java projects, would be nice to have a definitive answer and an example setup.

**Comment by kchodorow** [comment]
I'm not sure why I closed this via that commit, reopening. Nothing has happened with the @foo//jar:transitive idea yet, but I still think it should.

**Comment by spierce7** [comment]
@kchodorow So I'm assuming that pulling in transitive maven dependencies is something that still isn't working?

**Comment by kchodorow** [comment]
Er, I attempted to make it clear that it does not.  Feel free to send a patch if you have any ideas for improving the phrasing.

**Comment by damienmg** [comment]
@kchodorow: Can this one be closed now?

**Comment by kchodorow** [comment]
Yeah, I think at this point people can add this themselves with skylark repositories and/or generate it with generate_workspace.

**Comment by vuryleo** [comment]
If you already have `generate_workspace`, why not integrate it into bazel and make transitive maven depen dencies becomes true?

**Comment by wstrange** [comment]
Even though this issue is closed, I just wanted to add that maven transitive dependencies is a real pain point for people thinking of converting to bazel. 

The generate_workspace util should at least be packaged so people don't need to build from source.

**Comment by davidsantiago** [comment]
I agree. This issue has single handedly ruled out the use of Bazel on recent projects for me. I understand the difficulty of it and am totally into the Bazel Way, but the management of the scores of maven jars I needed was just too much work. Especially when combined with the work needed to get various tools working nicely with Bazel that have ready made plugins in other build systems. (Including, disappointingly, a couple of google projects).

**Comment by ScottPierce** [comment]
If bazel was able to pull and resolve transitive maven repositories in a similar way to gradle this would make bazel a far more compelling build system for me.

**Comment by pauldraper** [comment]
Agreed.

(1) Checking in script outputs is not very DRY. A build system that requires this is a weird build indeed.

(2) As pointed out earlier, `java_library` manages to transitively add dependencies to classpath; we just need that for remote jars.

(3)

> make specifying a sha1 for every dependency awkward

That is *exactly* what is required today. With a transitive rule, you could specify exactly one SHA hash for the entire resolved result.

I don't seem a credible argument for why transitivity is bad.

**Comment by bluecmd** [comment]
> That is exactly what is required today. With a transitive rule, you could specify exactly one SHA hash for the entire resolved result.

The way I read the comment was not that it was awkward to do it, but that way proposed would make it awkward.

Doing transitive dependency resolution automatic is a sure way to make dependencies you have no idea that you're using end up in your build artifacts, something that scares me.

FWIW, I'm using `generate_workspace` and it works pretty well. Not the most user-friendly thing in the neighborhood but gets the job done.

**Comment by pauldraper** [comment]
> dependencies you have no idea that you're using end up in your build artifacts

And you generally don't need to know. Encapsulation + abstraction. You don't care how a library is implemented, just as long as it works. True, there are times when that is relevant, and you can check.

Regardless, running a program (generate_workspace) to generate a program (WORKSPACE) to generate a program (jar) is odd. A build tool should be able to combine those steps through automation.

**Comment by bluecmd** [comment]
> And you generally don't need to know. Encapsulation + abstraction. You don't care how a library is implemented, just as long as it works.

I can think of plenty of cases where you would care. I care what licenses we use, and besides sometimes Maven dependencies needs tuning ("I'm not using MySQL, and this package depends on it for a single mode of operation"*). A developer simply +1'ing the version and causes a new dependency to get pulled in is bad. Sure, we could have presubmits and what not, but those would be Maven specific. We have more stuff than Java, and having all external dependencies handled in the same way works very well.

> A build tool should be able to combine those steps through automation.

I agree that it would be a useful feature for Bazel to have to reach critical mass, a lot of projects don't care. For my case however, one of the selling points for Bazel was that it does not do these magic resolutions and we're in absolute control over what gets pulled in.

\* One could argue it is the library author's responsibility to package it in a modular way, but that's a pipe dream.

**Comment by azymnis** [comment]
FYI here is one possible approach that seems to work ok: https://github.com/johnynek/bazel-deps

**Comment by kant111** [comment]
This is a deal breaker for me as well. why not support both approaches i.e specifying transitive dependencies explicitly vs automatically pulling the transitive dependencies and let the users decide on whatever they want to do instead of enforcing it? 

A project can have a bunch of dependencies which in turn can have more dependencies  which in turn can have more dependencies and it goes on..Now If I have 100's of jar how can we expect a user to verify every single one of them? Moreover how does a user know what to verify besides resolving the version conflicts ? If version conflicts happens so often how come Maven and gradle where able to go so far? For people who want to verify 100's of jars or whatever the number is It's them you actually need a tool like `generate_workspace` (not all regular users) so they can verify when there is a version conflict and be able to fix it by specifying explicitly for those jars that are in conflict.

In short, this is just not user friendly and so the price is huge and again there it appears to me that there is no good reason to not support both approaches. 

I also don't understand why this ticket is closed without proper consensus. This is not open source. This is Google source I guess :)

**Comment by pcj** [comment]
Here's the approach I took when writing [rules_maven](https://github.com/pubref/rules_maven) that I think satisfies many of the pain points echoed here (as I too shared them).

1. Avoid having the user to invoke some other build step, it should be automated as per @pauldraper's comment.

1. Don't re-write code for transitive dependency resolution when it's already been done pretty well by gradle. [So, just use it (gradle) as an internal tool (implementation detail)].

1. Give the user a way to lock down the expected sha1 values for resolved jars in the form `[SHA1:GROUP:NAME:VERSION]`.

If you write:

```python
maven_repository(
    name = 'grpc_java',
    deps = [
        'io.grpc:grpc-all:1.2.0',
    ],
    force = [
        'com.google.guava:guava:19.0',
    ],
)
```

You get:

```python
maven_repository(
    name = 'grpc_java',
    deps = [
        'io.grpc:grpc-all:1.2.0',
    ],
    force = [
        'com.google.guava:guava:19.0',
    ],
    transitive_deps = [
        '171da91494a1391aef13b88bd7302b29edb8d3b3:com.google.auth:google-auth-library-credentials:0.4.0',
        '5871fb60dc68d67da54a663c3fd636a10a532948:com.google.code.findbugs:jsr305:3.0.0',
        '751f548c85fa49f330cecbb1875893f971b33c4e:com.google.code.gson:gson:2.7',
        '3624d81fca4e93c67f43bafc222b06e1b1e3b260:com.google.errorprone:error_prone_annotations:2.0.11',
        '6ce200f6b23222af3d8abb6b6459e6c44f4bb0e9:com.google.guava:guava:19.0',
        'a2e145e7a7567c6372738f5c5a6f3ba6407ac354:com.google.instrumentation:instrumentation-api:0.3.0',
        '357e60f95cebb87c72151e49ba1f570d899734f8:com.google.protobuf.nano:protobuf-javanano:3.0.0-alpha-5',
        '62ccf171a106ff6791507f2d5364c275f9a3131d:com.google.protobuf:protobuf-java:3.2.0',
        '25ec9bf5bb119978c514b46f568bed896c5a4a8b:com.google.protobuf:protobuf-java-util:3.2.0',
        '59b5b9c6e1a3054696d23492f888c1f8b583f5fc:com.google.protobuf:protobuf-lite:3.0.1',
        '4de2b4ed3445c37ec1720a7d214712e845a24636:com.squareup.okhttp:okhttp:2.5.0',
        '98476622f10715998eacf9240d6b479f12c66143:com.squareup.okio:okio:1.6.0',
        'f32006a1245dfa2d68bf92a1b4cc01831889c95b:io.grpc:grpc-all:1.2.0',
        '540e20dffcaf91d4a18cb57637f2bfdfc3bb3676:io.grpc:grpc-auth:1.2.0',
        '1932db544cbb427bc18f902c7ebbb3f7e44991df:io.grpc:grpc-context:1.2.0',
        'f12a213e2b59a0615df2cc9bed35dc15fd2fee37:io.grpc:grpc-core:1.2.0',
        'e2682d2dc052898f87433e7a6d03d104ef98df74:io.grpc:grpc-netty:1.2.0',
        'fc17023227651ccf369570588b4516a96ca32fed:io.grpc:grpc-okhttp:1.2.0',
        '2676852d2dbd20155d9b1a940a456eae5b7445f0:io.grpc:grpc-protobuf:1.2.0',
        '5ecc9a6e23e11694f99fad3e66e12cc3ce3b93b4:io.grpc:grpc-protobuf-lite:1.2.0',
        '58d8d4ec67fb1c929802450f23f5d5a3a036d3ac:io.grpc:grpc-protobuf-nano:1.2.0',
        '964dda53b3085bfd17c7aaf51495f9efc8bda36c:io.grpc:grpc-stub:1.2.0',
        '43292c2622e340a0d07178c341ca3bdf3d662034:io.netty:netty-buffer:4.1.8.Final',
        '1bd0a2d032e5c7fc3f42c1b483d0f4c57eb516a3:io.netty:netty-codec:4.1.8.Final',
        '1e88617c4a6c88da7e86fdbbd9494d22a250c879:io.netty:netty-codec-http:4.1.8.Final',
        '105a99ee5767463370ccc3d2e16800bd99f5648e:io.netty:netty-codec-http2:4.1.8.Final',
        '7f7c5f5b154646d7c571f8ca944fb813f71b1d51:io.netty:netty-codec-socks:4.1.8.Final',
        'ee62c80318413d2375d145e51e48d7d35c901324:io.netty:netty-common:4.1.8.Final',
        'db01139bfb11afd009a695eef55b43bbf22c4ef5:io.netty:netty-handler:4.1.8.Final',
        'c4d22e8b9071a0ea8d75766d869496c32648a758:io.netty:netty-handler-proxy:4.1.8.Final',
        '2e116cdd5edc01b2305072b1dbbd17c0595dbfef:io.netty:netty-resolver:4.1.8.Final',
        '905b5dadce881c9824b3039c0df36dabbb7b6a07:io.netty:netty-transport:4.1.8.Final',
    ],
)

**Comment by jin** [comment]
An update: https://github.com/bazelbuild/rules_jvm_external is the officially maintained ruleset by the Bazel team to fetch and resolve artifacts transitively.
